browser/brave_wallet/asset_discovery_manager_unittest.cc

```diff
@@ -0,0 +1,1201 @@
+        [&, responses, expected_from_blocks,
+         expected_to_blocks](const network::ResourceRequest& request) {
+          base::StringPiece request_string(request.request_body->elements()
+                                               ->at(0)
+                                               .As<network::DataElementBytes>()
+                                               .AsStringPiece());
+          const auto expected_from_block =
+              expected_from_blocks.find(request.url.spec())->second;
+          const auto expected_to_block =
+              expected_to_blocks.find(request.url.spec())->second;
+
+          EXPECT_NE(request_string.find(base::StringPrintf(
+                        "\"fromBlock\":\"%s\"", expected_from_block.c_str())),
+                    std::string::npos);
+          EXPECT_NE(request_string.find(base::StringPrintf(
+                        "\"toBlock\":\"%s\"", expected_to_block.c_str())),
+                    std::string::npos);
+          url_loader_factory_.ClearResponses();
+          for (const auto& kv : responses) {
+            url_loader_factory_.AddResponse(kv.first, kv.second);
+          }
+          return;
+        }));
+  }
+
+  void TestDiscoverAssets(
+      const std::string& chain_id,
+      mojom::CoinType coin,
+      const std::vector<std::string>& account_addresses,
+      const std::vector<std::string>& expected_token_contract_addresses,
+      mojom::ProviderError expected_error,
+      const std::string& expected_error_message,
+      const std::string& expected_next_asset_discovery_from_block) {
+    // Set remaining chains to 1 in order since this value needs to end at
+    // 0 by the end of the DiscoverAssets call in order to trigger the
+    // event and it will not be set by an outer
+    // DiscoverAssetsOnAllSupportedChains* call in this unit test.
+    asset_discovery_manager_->remaining_chains_ = 1;
+    base::RunLoop run_loop;
+    asset_discovery_manager_->DiscoverAssets(
+        chain_id, mojom::CoinType::ETH, account_addresses, false,
+        kEthereumBlockTagEarliest, kEthereumBlockTagLatest,
+        GetAssetDiscoverySupportedChainsForTest(),
+        base::BindLambdaForTesting(
+            [&](const std::string& chain_id,
+                const std::vector<mojom::BlockchainTokenPtr> discovered_assets,
+                mojom::ProviderError error, const std::string& error_message) {
+              EXPECT_EQ(chain_id, chain_id);
+              EXPECT_EQ(expected_error, error);
+              EXPECT_EQ(expected_error_message, error_message);
+              ASSERT_EQ(expected_token_contract_addresses.size(),
+                        discovered_assets.size());
+              for (size_t i = 0; i < discovered_assets.size(); i++) {
+                EXPECT_EQ(expected_token_contract_addresses[i],
+                          discovered_assets[i]->contract_address);
+              }
+            }));
+    run_loop.RunUntilIdle();
```

Comment:
Is this mainly for OnDiscoverAssetsCompleted event or any other things?
If it's just OnDiscoverAssetsCompleted event, I would suggest in the Observer class, have something like
```
// in this function:
WaitForOnDiscoverAssetsCompleted(expected_token_contract_addresses); 

// in observer:
void WaitForOnDiscoverAssetsCompleted(const std::vector<std::string>& addresses) {
  expected_contract_addresses_ = addresses;
  run_loop_asset_discovery_ = std::make_unique<base::RunLoop>();
  run_loop_asset_discovery_->Run(); 
}

void OnDiscoverAssetsCompleted(
      std::vector<mojom::BlockchainTokenPtr> discovered_assets) override {
  // TODO: check if addresses in discovered_assets matches expected_contract_addresses_.
  run_loop_asset_discovery_->Quit();
}
```

What you have below can be replaced by simply calling WaitForOnDiscoverAssetsCompleted.
```
run_loop.RunUntilIdle();
wallet_service_observer_->TestOnDiscoverAssetsCompletedResults(
        std::move(expected_token_contract_addresses));
    wallet_service_observer_->Reset();
```

Reply:
Hmm.. I tried this a bit but ran into some complexity.  The main thing is that some code paths will not emit an event -- namely when this is run in the background when new accounts are added (`AssetDiscoveryManager::DiscoverAssetsOnAllSupportedChainsAccountsAdded`).  So that test would hang

Reply:
I don't understand, in what cases there won't be events, at least it should always have the event for where you are doing `TestOnDiscoverAssetsCompletedResults`.

Reply:
And in cases if you aren't expecting an event, you wouldn't even write this "WaitForxxxx".

Reply:
run_loop.RunUntilIdle(); is vague, we should try to do run loop and quit as expected to make sure the event we want really happens.

Reply:
Sounds good. I was using TestOnDiscoverAssetsCompletedResults to test both cases (event and no event), and was thinking of WaitForOnDiscoverAssetsCompleted as a replacement for that.

I'll do as you suggested, and will create a different way to test when no event is fired.